// @linagora/rabbitmq-client/dist/index.js
// src/client.ts
import amqp from "amqplib";

// src/logger.ts
import pino from "pino";
var createLogger = (parent) => {
  if (parent) {
    return parent.child({ service: "rabbitmq-client" });
  }
  return pino({ name: "rabbitmq-client" });
};

// src/client.ts
var DEFAULT_PREFETCH = 10;
var MAX_PUBLISH_RETRY_DELAY_MS = 6e4;
var DEFAULTS = {
  maxRetries: 3,
  retryDelay: 1e3,
  connectionRetryDelay: 5e3,
  initMaxAttempts: 5,
  publishMaxAttempts: 5,
  prefetch: DEFAULT_PREFETCH,
  closeTimeout: 5e3
};
var RabbitMQClient = class {
  connection = null;
  channel = null;
  connected = false;
  subscriptions = [];
  options;
  logger;
  hooks;
  initializationPromise = null;
  reconnectionPromise = null;
  assertedExchanges = /* @__PURE__ */ new Set();
  consumerTags = /* @__PURE__ */ new Map();
  inflightCount = 0;
  drainResolve = null;
  constructor(options) {
    this.options = {
      url: options.url,
      maxRetries: options.maxRetries ?? DEFAULTS.maxRetries,
      retryDelay: options.retryDelay ?? DEFAULTS.retryDelay,
      connectionRetryDelay: options.connectionRetryDelay ?? DEFAULTS.connectionRetryDelay,
      initMaxAttempts: options.initMaxAttempts ?? DEFAULTS.initMaxAttempts,
      publishMaxAttempts: options.publishMaxAttempts ?? DEFAULTS.publishMaxAttempts,
      prefetch: options.prefetch ?? DEFAULTS.prefetch,
      closeTimeout: options.closeTimeout ?? DEFAULTS.closeTimeout
    };
    this.logger = createLogger(options.logger);
    this.hooks = options.hooks ?? {};
  }
  /**
   * Opens a connection and creates a confirm channel. Retries up to
   * `initMaxAttempts` times, then throws. Idempotent and safe to call
   * concurrently — duplicate calls share the same in-flight promise.
   */
  async init() {
    if (this.connection && this.connected) {
      return;
    }
    if (this.initializationPromise) {
      return this.initializationPromise;
    }
    this.initializationPromise = this.doConnect(this.options.initMaxAttempts);
    this.initializationPromise.catch(() => void 0);
    try {
      await this.initializationPromise;
    } finally {
      this.initializationPromise = null;
    }
  }
  async doConnect(maxAttempts) {
    this.assertedExchanges.clear();
    this.consumerTags.clear();
    let attempts = 0;
    while (!this.connected) {
      try {
        this.connection = await amqp.connect(this.options.url);
        this.logger.info("Connected to server");
        this.channel = await this.connection.createConfirmChannel();
        this.logger.info("Confirm channel created");
        await this.channel.prefetch(this.options.prefetch);
        this.logger.info({ prefetch: this.options.prefetch }, "Channel prefetch set");
        this.connection.on("error", (error) => {
          this.connected = false;
          this.logger.error({ error }, "Connection error");
        });
        this.connection.on("close", () => {
          this.connected = false;
          this.logger.warn("Connection closed");
          this.reconnectWithRetry();
        });
        this.channel.on("error", (error) => {
          this.logger.error({ error }, "Channel error");
          this.handleChannelFailure();
        });
        this.channel.on("close", () => {
          this.logger.warn("Channel closed");
          this.handleChannelFailure();
        });
        this.connected = true;
        this.logger.info("Client initialized successfully");
      } catch (error) {
        attempts++;
        this.connected = false;
        if (maxAttempts !== void 0 && attempts >= maxAttempts) {
          this.logger.error({ error, attempts, maxAttempts }, "Connection failed after maximum attempts");
          throw new Error(
            `Failed to connect to RabbitMQ after ${attempts} attempts. Check RABBITMQ_URL configuration and RabbitMQ server availability.`
          );
        }
        this.logger.warn(
          {
            error,
            attempt: attempts,
            maxAttempts: maxAttempts ?? "unlimited",
            retryDelayMs: this.options.connectionRetryDelay
          },
          "Connection attempt failed, retrying..."
        );
        await this.sleep(this.options.connectionRetryDelay);
      }
    }
  }
  handleChannelFailure() {
    if (this.connected) {
      this.connected = false;
      this.reconnectWithRetry();
    }
  }
  /**
   * Returns whether the client currently has an active connection and channel.
   */
  isConnected() {
    return this.connected;
  }
  /**
   * Publishes a JSON message to a topic exchange with publisher confirms.
   * Retries with exponential backoff (capped at 60 s), forcing a reconnect
   * on each failure. Throws after `publishMaxAttempts` exhausted.
   */
  async publish(exchange, routingKey, message, options) {
    let attempts = 0;
    const maxAttempts = options?.maxAttempts ?? this.options.publishMaxAttempts;
    const baseDelay = this.options.connectionRetryDelay;
    const content = Buffer.from(JSON.stringify(message));
    while (attempts < maxAttempts) {
      try {
        if (!this.connected) {
          await this.reconnectWithRetry();
        }
        if (!this.channel) {
          throw new Error("Channel not available");
        }
        if (!this.assertedExchanges.has(exchange)) {
          await this.channel.assertExchange(exchange, "topic", { durable: true });
          this.assertedExchanges.add(exchange);
        }
        this.channel.publish(exchange, routingKey, content, {
          persistent: true,
          headers: options?.headers,
          correlationId: options?.correlationId,
          messageId: options?.messageId,
          expiration: options?.expiration
        });
        await this.channel.waitForConfirms();
        if (attempts > 0) {
          this.logger.info({ exchange, routingKey, messageSize: content.length, attempts }, "Published message after retries");
        } else {
          this.logger.info({ exchange, routingKey, messageSize: content.length }, "Published message");
        }
        this.logger.debug({ exchange, routingKey, payload: message }, "Published message payload");
        this.callHook(this.hooks.onPublish, { exchange, routingKey, attempts: attempts + 1 });
        return;
      } catch (error) {
        attempts++;
        this.connected = false;
        if (attempts >= maxAttempts) {
          this.logger.error(
            { error, exchange, routingKey, attempts, maxAttempts },
            "Publish failed after max attempts"
          );
          throw new Error(
            `Failed to publish to ${exchange}/${routingKey} after ${attempts} attempts: ${error instanceof Error ? error.message : String(error)}`
          );
        }
        const retryDelay = Math.min(
          baseDelay * Math.pow(2, attempts - 1),
          MAX_PUBLISH_RETRY_DELAY_MS
        );
        this.logger.warn(
          { error, exchange, routingKey, attempt: attempts, maxAttempts, retryDelayMs: retryDelay },
          "Publish attempt failed, retrying"
        );
        await this.sleep(retryDelay);
      }
    }
  }
  /**
   * Gracefully shuts down the client. Waits up to `closeTimeout` ms for
   * in-flight message handlers to finish before closing the channel and
   * connection. Pass `false` to preserve subscriptions for a later
   * `init()` / reconnect cycle.
   */
  async close(clearSubscriptions = true) {
    try {
      if (this.inflightCount > 0) {
        this.logger.info({ inflightCount: this.inflightCount }, "Waiting for in-flight messages to drain");
        await this.waitForDrain(this.options.closeTimeout);
      }
      await this.channel?.close();
      await this.connection?.close();
      this.connection = null;
      this.channel = null;
      this.connected = false;
      this.initializationPromise = null;
      this.reconnectionPromise = null;
      this.assertedExchanges.clear();
      this.consumerTags.clear();
      if (clearSubscriptions) {
        this.subscriptions = [];
      }
      this.logger.info("Connection closed");
    } catch (error) {
      this.logger.error({ error }, "Error closing connection");
      throw error;
    }
  }
  async reconnectWithRetry() {
    if (this.reconnectionPromise) {
      return this.reconnectionPromise;
    }
    this.connected = false;
    this.logger.info("Starting reconnection...");
    this.reconnectionPromise = this.doReconnect();
    this.reconnectionPromise.catch(() => void 0);
    try {
      await this.reconnectionPromise;
    } finally {
      this.reconnectionPromise = null;
    }
  }
  async doReconnect() {
    await this.doConnect();
    await this.resubscribeAll();
  }
  async resubscribeAll() {
    if (this.subscriptions.length === 0) {
      return;
    }
    this.logger.info({ count: this.subscriptions.length }, "Re-establishing subscriptions");
    const subs = [...this.subscriptions];
    const results = await Promise.allSettled(
      subs.map(async (sub) => {
        await this.setupSubscription(sub);
        return sub.queue;
      })
    );
    const succeeded = [];
    const failed = [];
    results.forEach((result, index) => {
      if (result.status === "fulfilled") {
        succeeded.push(result.value);
      } else {
        const queue = subs[index].queue;
        failed.push(queue);
        this.logger.error({ error: result.reason, queue }, "Failed to re-subscribe to queue");
      }
    });
    if (succeeded.length > 0) {
      this.logger.info({ count: succeeded.length, queues: succeeded }, "Successfully re-subscribed to queues");
    }
    if (failed.length > 0) {
      this.logger.warn({ count: failed.length, queues: failed }, "Some subscriptions failed to restore");
    }
    this.callHook(this.hooks.onReconnect, { subscriptionsRestored: succeeded.length, subscriptionsFailed: failed.length });
  }
  /**
   * Subscribes to a queue with automatic DLQ infrastructure setup.
   *
   * Pass `options.queueArguments` to override the default quorum-queue
   * arguments (merged with the DLQ wiring defaults).
   */
  async subscribe(exchange, routingKey, queue, handler, options) {
    if (!this.connected || !this.channel) {
      throw new Error("RabbitMQ client not connected. Call init() first.");
    }
    const sub = { exchange, routingKey, queue, handler, options };
    const existingIndex = this.subscriptions.findIndex((s) => s.queue === queue);
    if (existingIndex === -1) {
      this.subscriptions.push(sub);
    } else {
      this.subscriptions[existingIndex] = sub;
    }
    await this.setupSubscription(sub);
  }
  /**
   * Cancels a queue subscription and removes it from the restoration list.
   */
  async unsubscribe(queue) {
    const tag = this.consumerTags.get(queue);
    if (tag && this.channel) {
      await this.channel.cancel(tag);
    }
    this.consumerTags.delete(queue);
    this.subscriptions = this.subscriptions.filter((s) => s.queue !== queue);
    this.logger.info({ queue }, "Unsubscribed from queue");
  }
  async setupSubscription(sub) {
    if (!this.channel) {
      throw new Error("Channel not available");
    }
    const { exchange, routingKey, queue, handler, options } = sub;
    const dlxExchange = `${exchange}.dlx`;
    const dlqQueue = `${queue}.dlq`;
    const dlqRoutingKey = `${routingKey}.dead`;
    await this.channel.assertExchange(dlxExchange, "topic", { durable: true });
    await this.channel.assertQueue(dlqQueue, { durable: true });
    await this.channel.bindQueue(dlqQueue, dlxExchange, dlqRoutingKey);
    if (!this.assertedExchanges.has(exchange)) {
      await this.channel.assertExchange(exchange, "topic", { durable: true });
      this.assertedExchanges.add(exchange);
    }
    const queueType = options?.queueArguments?.["x-queue-type"] ?? "quorum";
    const queueArgs = {
      "x-queue-type": queueType,
      "x-overflow": "reject-publish"
    };
    if (queueType === "quorum") {
      queueArgs["x-dead-letter-strategy"] = "at-least-once";
    }
    await this.channel.assertQueue(queue, {
      durable: true,
      deadLetterExchange: dlxExchange,
      deadLetterRoutingKey: dlqRoutingKey,
      arguments: { ...queueArgs, ...options?.queueArguments }
    });
    await this.channel.bindQueue(queue, exchange, routingKey);
    const { consumerTag } = await this.channel.consume(
      queue,
      (message) => {
        if (message) {
          this.handleWithRetry(message, handler).catch((error) => {
            this.logger.error({ error }, "Unhandled error in message handler");
          });
        }
      },
      { noAck: false }
    );
    this.consumerTags.set(queue, consumerTag);
    this.logger.info({ exchange, routingKey, queue }, "Subscribed to queue");
  }
  async handleWithRetry(message, handler) {
    const channel = this.channel;
    if (!channel) {
      this.logger.warn("Cannot process message -- channel unavailable");
      return;
    }
    this.inflightCount++;
    try {
      const startTime = Date.now();
      let attempts = 0;
      const routingKey = message.fields.routingKey;
      const exchange = message.fields.exchange;
      let content;
      try {
        content = JSON.parse(message.content.toString());
      } catch (parseError) {
        const rawContent = message.content.toString();
        const rawPreview = rawContent.substring(0, 100);
        this.logger.error(
          {
            error: parseError,
            exchange,
            routingKey,
            rawContentPreview: rawPreview + (rawContent.length > 100 ? "..." : "")
          },
          "Failed to parse message JSON, sending to DLQ"
        );
        channel.nack(message, false, false);
        this.callHook(this.hooks.onMessageDlq, { exchange, routingKey, duration: 0, reason: "invalid_json" });
        return;
      }
      this.logger.debug({ exchange, routingKey, payload: content }, "Message received, processing");
      while (attempts < this.options.maxRetries) {
        try {
          await handler(content);
          const duration2 = Date.now() - startTime;
          this.logger.info(
            { exchange, routingKey, duration: duration2, attempts: attempts + 1 },
            "Message processed successfully"
          );
          channel.ack(message);
          this.callHook(this.hooks.onMessageProcessed, { exchange, routingKey, duration: duration2, attempts: attempts + 1 });
          return;
        } catch (error) {
          attempts++;
          this.logger.error(
            {
              error: error instanceof Error ? error.message : error,
              stack: error instanceof Error ? error.stack : void 0,
              exchange,
              routingKey,
              attempt: attempts,
              maxRetries: this.options.maxRetries
            },
            "Handler failed"
          );
          if (attempts < this.options.maxRetries) {
            await this.sleep(this.options.retryDelay);
          }
        }
      }
      const duration = Date.now() - startTime;
      this.logger.error(
        { exchange, routingKey, maxRetries: this.options.maxRetries, duration },
        "Message failed after max retries, sending to DLQ"
      );
      channel.nack(message, false, false);
      this.callHook(this.hooks.onMessageDlq, { exchange, routingKey, duration, reason: "max_retries_exhausted" });
    } finally {
      this.inflightCount--;
      if (this.inflightCount === 0 && this.drainResolve) {
        this.drainResolve();
      }
    }
  }
  /**
   * Lightweight liveness probe: creates and immediately deletes a temporary
   * exclusive queue. Returns `false` when disconnected or on broker error.
   */
  async checkHealth() {
    if (!this.connected || !this.channel) {
      return false;
    }
    try {
      const { queue } = await this.channel.assertQueue("", {
        exclusive: true,
        autoDelete: true
      });
      await this.channel.deleteQueue(queue);
      return true;
    } catch (error) {
      this.logger.warn({ error }, "Health check failed");
      return false;
    }
  }
  waitForDrain(timeout) {
    if (this.inflightCount === 0) return Promise.resolve();
    return new Promise((resolve) => {
      const timer = setTimeout(() => {
        this.logger.warn(
          { inflightCount: this.inflightCount },
          "Close timeout reached with messages still in flight"
        );
        this.drainResolve = null;
        resolve();
      }, timeout);
      this.drainResolve = () => {
        clearTimeout(timer);
        this.drainResolve = null;
        resolve();
      };
    });
  }
  /** Invokes a hook callback, swallowing errors so hooks never break message flow. */
  callHook(hook, info) {
    try {
      hook?.(info);
    } catch (error) {
      this.logger.debug({ error }, "Observability hook error");
    }
  }
  sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
};
export {
  RabbitMQClient
};
