// @linagora/rabbitmq-client/dist/testing/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/testing/index.ts
var testing_exports = {};
__export(testing_exports, {
  MockRabbitMQChannel: () => MockRabbitMQChannel,
  MockRabbitMQConnection: () => MockRabbitMQConnection,
  createMockAmqplib: () => createMockAmqplib
});
module.exports = __toCommonJS(testing_exports);
var import_events = require("events");
function mockFn() {
  let impl = null;
  const calls = [];
  const fn = ((...args) => {
    calls.push(args);
    return impl ? impl(...args) : void 0;
  });
  fn.mock = { calls };
  fn.mockResolvedValue = (val) => {
    impl = () => Promise.resolve(val);
    return fn;
  };
  fn.mockImplementation = (newImpl) => {
    impl = newImpl;
    return fn;
  };
  return fn;
}
var MockRabbitMQChannel = class extends import_events.EventEmitter {
  _publishedMessages = [];
  _consumeCallback = null;
  isClosed = false;
  // Stubs
  assertExchange = mockFn().mockResolvedValue({});
  assertQueue = mockFn().mockResolvedValue({ queue: "" });
  bindQueue = mockFn().mockResolvedValue({});
  ack = mockFn();
  nack = mockFn();
  close = mockFn().mockResolvedValue(void 0);
  waitForConfirms = mockFn().mockResolvedValue(void 0);
  prefetch = mockFn().mockResolvedValue(void 0);
  deleteQueue = mockFn().mockResolvedValue({});
  cancel = mockFn().mockResolvedValue(void 0);
  /**
   * Tracks all published messages and stores them for later inspection.
   */
  publish(exchange, routingKey, content, options) {
    let message;
    try {
      message = JSON.parse(content.toString());
    } catch {
      message = content.toString();
    }
    this._publishedMessages.push({ exchange, routingKey, message, options });
    return true;
  }
  _consumerCounter = 0;
  /**
   * Stores the consumer callback for use with simulateMessage.
   * Returns a unique consumer tag per call.
   */
  consume(_queue, callback) {
    this._consumeCallback = callback;
    const consumerTag = `mock-consumer-${++this._consumerCounter}`;
    return Promise.resolve({ consumerTag });
  }
  // ---------------------------------------------------------------------------
  // Simulation helpers
  // ---------------------------------------------------------------------------
  /**
   * Delivers a valid JSON message to the registered consume callback.
   */
  simulateMessage(content) {
    this._deliver(Buffer.from(JSON.stringify(content)));
  }
  /**
   * Delivers a message with content that is not valid JSON.
   */
  simulateInvalidJsonMessage() {
    this._deliver(Buffer.from("not valid json {"));
  }
  _deliver(content) {
    if (!this._consumeCallback) return;
    this._consumeCallback({
      content,
      fields: { deliveryTag: 1, redelivered: false, exchange: "", routingKey: "", consumerTag: "mock-consumer" },
      properties: { headers: {} }
    });
  }
  /** Emits a close event and marks the channel as closed. */
  simulateClose() {
    this.isClosed = true;
    this.emit("close");
  }
  /** Emits an error event. */
  simulateError(error) {
    this.emit("error", error);
  }
  // ---------------------------------------------------------------------------
  // Inspection helpers
  // ---------------------------------------------------------------------------
  getPublishedMessages() {
    return [...this._publishedMessages];
  }
  clearMessages() {
    this._publishedMessages = [];
  }
  /** Resets all state back to initial values. */
  reset() {
    this._publishedMessages = [];
    this._consumeCallback = null;
    this._consumerCounter = 0;
    this.isClosed = false;
    this.removeAllListeners();
  }
};
var MockRabbitMQConnection = class extends import_events.EventEmitter {
  _channels = [];
  isConnected = true;
  constructor() {
    super();
    this._channels.push(new MockRabbitMQChannel());
  }
  /**
   * Returns the most recently created channel.
   */
  get channel() {
    return this._channels[this._channels.length - 1];
  }
  /**
   * Returns all channels created so far (including the initial one).
   */
  get channels() {
    return [...this._channels];
  }
  get channelCount() {
    return this._channels.length;
  }
  /**
   * Creates a new channel (simulates a reconnection / new confirm-channel).
   */
  createConfirmChannel() {
    const ch = new MockRabbitMQChannel();
    this._channels.push(ch);
    return Promise.resolve(ch);
  }
  close() {
    this.isConnected = false;
    return Promise.resolve();
  }
  /** Emits a close event and marks the connection as disconnected. */
  simulateClose() {
    this.isConnected = false;
    this.emit("close");
  }
  /** Emits an error event. */
  simulateError(error) {
    this.emit("error", error);
  }
  /** Resets to the initial connected state with a single fresh channel. */
  reset() {
    this._channels = [new MockRabbitMQChannel()];
    this.isConnected = true;
    this.removeAllListeners();
  }
};
function createMockAmqplib() {
  const mockConnection = new MockRabbitMQConnection();
  const amqpMock = {
    connect: mockFn().mockResolvedValue(mockConnection)
  };
  return { mockConnection, amqpMock };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  MockRabbitMQChannel,
  MockRabbitMQConnection,
  createMockAmqplib
});
